torch/distributed/_spmd/graph_optimization.py

```diff
@@ -0,0 +1,408 @@
+    node.meta["tensor_meta"] = _create_meta_tensor_meta(fake_tensor_mode, new_meta_val)
+    return node
+
+
+def _scatter_wait_result(
+    gm: fx.GraphModule,
+    fused_comm_block: CommBlock,
+    comm_blocks: List[CommBlock],
+    node_indices: Dict[fx.Node, int],
+) -> None:
+    last_wait_node_idx = 0
+    for node in gm.graph.nodes:
+        if node == fused_comm_block.comm_node:
+            break
+        last_wait_node_idx = max(
+            node_indices.get(node, last_wait_node_idx), last_wait_node_idx
+        )
+
+    fused_comm_node = fused_comm_block.comm_node
+    fused_wait_node = fused_comm_block.wait_nodes[0]
+
+    with gm.graph.inserting_after(fused_wait_node):
+        split_node = gm.graph.call_function(
+            aten.split, (fused_wait_node, [cb.shape.numel() for cb in comm_blocks])
+        )
+
+    need_sort_nodes = []
+    last_split_reshape_node = split_node
+    with gm.graph.inserting_after(split_node):
+        for idx, comm_block in enumerate(comm_blocks):
+            orig_wait = comm_block.wait_nodes[0]
+            nodes = collections.deque(list(orig_wait.users))
+            split_idx_node = gm.graph.call_function(operator.getitem, (split_node, idx))
+            with gm.graph.inserting_after(split_idx_node):
+                wait_output_node = gm.graph.call_function(
+                    aten.reshape, (split_idx_node, comm_block.shape)
+                )
+            while nodes:
+                user_node = nodes.popleft()
+                if not isinstance(user_node, fx.Node):
+                    continue
+                if node_indices[user_node] < last_wait_node_idx:
+                    need_sort_nodes.append(user_node)
+                    nodes.extend(list(user_node.users))
+
+            gm.graph.node_replace_all_uses_with(orig_wait, wait_output_node)
+        if last_split_reshape_node == split_node:
+            last_split_reshape_node = wait_output_node
+
+    need_sort_nodes = sorted(need_sort_nodes, key=lambda node: node_indices[node])
+    gm.graph.move_after(need_sort_nodes, last_split_reshape_node)
+
+    gm.graph.eliminate_dead_code()
+
+
+def _fuse_with_cat(
+    gm: fx.GraphModule,
+    comm_blocks: List[CommBlock],
+    node_indices: Dict[fx.Node, int],
+) -> fx.Node:
+    # Find the last input node.
+    last_input_node = comm_blocks[0].inputs[0]
+    last_input_index = -1
+    all_input_nodes = []
+    for comm_block in comm_blocks:
+        input_node = comm_block.inputs[0]
+        # If the input node is a clone, this is CommTensor based implementation.
+        if input_node.name.startswith("clone"):
+            input_node = input_node.args[0]
+        all_input_nodes.append(input_node)
+        index = node_indices[input_node]
+        if index >= last_input_index:
+            assert index != last_input_index
+            last_input_node = input_node
+            last_input_index = index
+
+    with gm.graph.inserting_after(last_input_node):
```

Comment:
curious, will other graph optimization passes move any of these input compute nodes around?

Reply:
Right now, no. Are you suggesting that we should move the input nodes earlier?

Reply:
I do not know. What if we fond an independent compute block (which contains the input) that we wanna move to an earlier place to overlap with other comms? Just a random thought.

Reply:
That's actually very possible I think. We should try if to squeeze more from the computation graph. Right now we just change the communication order.